sql: SUBSTRING of bit types not supported

Repro:
```
root@:26257/defaultdb> SELECT substring('0110'::bit(4), 1, 2);
ERROR: unknown signature: substring(bit, int, int)
```
EXPECTED: No error, PG gives this:
```
template1=# SELECT substring('0110'::bit(4), 1, 2);
 substring
-----------
 01
```

2+ hours in yet i still cannot get a clean build && passed test out of a freshly cloned repo following [this](https://wiki.crdb.io/wiki/spaces/CRDB/pages/181338446/Getting+and+building+from+source)  😢 

i used the docker image to build. first i got this:
```
Makefile:1506: bin/execgen_out.d: No such file or directory
```
i commented out the line that caused this and fire another build. it went for quite a while until I got this:

```
ERROR in ./src/views/jobs/jobTable.tsx
Module not found: Error: Can't resolve 'oss/src/components/pagination/pagination' in '/go/src/github.com/cockroachdb/cockroach/pkg/ui/src/views/jobs'
 @ ./src/views/jobs/jobTable.tsx 34:0-79 194:31-50
 @ ./src/views/jobs/index.tsx
 @ ./src/app.tsx
 @ ./src/index.tsx
 @ multi ./src/index.tsx
Makefile:1360: recipe for target 'pkg/ui/distccl/bindata.go' failed
```
I uncomment the previous line, tried again but got the same error. and somehow during this process `git status` refuse to work giving me this:

```
fatal: not a git repository: /go/src/github.com/cockroachdb/cockroach/.git/modules/c-deps/protobuf/modules/third_party/benchmark
fatal: 'git status --porcelain=2' failed in submodule c-deps/protobuf
```

lastly i run `./build/builder.sh make test` wishfully thinking it might be bogus. after about 20m it timed out with a lot of error logs:

```
panic: test timed out after 20m0s

goroutine 453006 [running]:
testing.(*M).startAlarm.func1()
	/usr/local/go/src/testing/testing.go:1377 +0xdf
created by time.goFunc
	/usr/local/go/src/time/sleep.go:168 +0x44
......
```

but the test seems not stopped, still driving my laptop's fan crazy...

can i get some help? it's really frustrating to start contributing like this...

--- a/docs/generated/sql/functions.md
+++ b/docs/generated/sql/functions.md
@@ -1045,6 +1045,10 @@ has no relationship with the commit order of concurrent transactions.</p>
 <tr><td><a name="strpos"></a><code>strpos(input: <a href="string.html">string</a>, find: <a href="string.html">string</a>) &rarr; <a href="int.html">int</a></code></td><td><span class="funcdesc"><p>Calculates the position where the string <code>find</code> begins in <code>input</code>.</p>
 <p>For example, <code>strpos('doggie', 'gie')</code> returns <code>4</code>.</p>
 </span></td></tr>
+<tr><td><a name="substr"></a><code>substr(input: <a href="bytes.html">bytes</a>, start_pos: <a href="int.html">int</a>) &rarr; <a href="bytes.html">bytes</a></code></td><td><span class="funcdesc"><p>Returns a byte subarray of <code>input</code> starting at <code>start_pos</code> (count starts at 1).</p>
+</span></td></tr>
+<tr><td><a name="substr"></a><code>substr(input: <a href="bytes.html">bytes</a>, start_pos: <a href="int.html">int</a>, length: <a href="int.html">int</a>) &rarr; <a href="bytes.html">bytes</a></code></td><td><span class="funcdesc"><p>Returns a byte subarray of <code>input</code> starting at <code>start_pos</code> (count starts at 1) and including up to <code>length</code> characters.</p>
+</span></td></tr>
 <tr><td><a name="substr"></a><code>substr(input: <a href="string.html">string</a>, regex: <a href="string.html">string</a>) &rarr; <a href="string.html">string</a></code></td><td><span class="funcdesc"><p>Returns a substring of <code>input</code> that matches the regular expression <code>regex</code>.</p>
 </span></td></tr>
 <tr><td><a name="substr"></a><code>substr(input: <a href="string.html">string</a>, regex: <a href="string.html">string</a>, escape_char: <a href="string.html">string</a>) &rarr; <a href="string.html">string</a></code></td><td><span class="funcdesc"><p>Returns a substring of <code>input</code> that matches the regular expression <code>regex</code> using <code>escape_char</code> as your escape character instead of <code>\</code>.</p>
@@ -1053,6 +1057,14 @@ has no relationship with the commit order of concurrent transactions.</p>
 </span></td></tr>
 <tr><td><a name="substr"></a><code>substr(input: <a href="string.html">string</a>, start_pos: <a href="int.html">int</a>, length: <a href="int.html">int</a>) &rarr; <a href="string.html">string</a></code></td><td><span class="funcdesc"><p>Returns a substring of <code>input</code> starting at <code>start_pos</code> (count starts at 1) and including up to <code>length</code> characters.</p>
 </span></td></tr>
+<tr><td><a name="substr"></a><code>substr(input: varbit, start_pos: <a href="int.html">int</a>) &rarr; varbit</code></td><td><span class="funcdesc"><p>Returns a bit subarray of <code>input</code> starting at <code>start_pos</code> (count starts at 1).</p>
+</span></td></tr>
+<tr><td><a name="substr"></a><code>substr(input: varbit, start_pos: <a href="int.html">int</a>, length: <a href="int.html">int</a>) &rarr; varbit</code></td><td><span class="funcdesc"><p>Returns a bit subarray of <code>input</code> starting at <code>start_pos</code> (count starts at 1) and including up to <code>length</code> characters.</p>
+</span></td></tr>
+<tr><td><a name="substring"></a><code>substring(input: <a href="bytes.html">bytes</a>, start_pos: <a href="int.html">int</a>) &rarr; <a href="bytes.html">bytes</a></code></td><td><span class="funcdesc"><p>Returns a byte subarray of <code>input</code> starting at <code>start_pos</code> (count starts at 1).</p>
+</span></td></tr>
+<tr><td><a name="substring"></a><code>substring(input: <a href="bytes.html">bytes</a>, start_pos: <a href="int.html">int</a>, length: <a href="int.html">int</a>) &rarr; <a href="bytes.html">bytes</a></code></td><td><span class="funcdesc"><p>Returns a byte subarray of <code>input</code> starting at <code>start_pos</code> (count starts at 1) and including up to <code>length</code> characters.</p>
+</span></td></tr>
 <tr><td><a name="substring"></a><code>substring(input: <a href="string.html">string</a>, regex: <a href="string.html">string</a>) &rarr; <a href="string.html">string</a></code></td><td><span class="funcdesc"><p>Returns a substring of <code>input</code> that matches the regular expression <code>regex</code>.</p>
 </span></td></tr>
 <tr><td><a name="substring"></a><code>substring(input: <a href="string.html">string</a>, regex: <a href="string.html">string</a>, escape_char: <a href="string.html">string</a>) &rarr; <a href="string.html">string</a></code></td><td><span class="funcdesc"><p>Returns a substring of <code>input</code> that matches the regular expression <code>regex</code> using <code>escape_char</code> as your escape character instead of <code>\</code>.</p>
@@ -1060,6 +1072,10 @@ has no relationship with the commit order of concurrent transactions.</p>
 <tr><td><a name="substring"></a><code>substring(input: <a href="string.html">string</a>, start_pos: <a href="int.html">int</a>) &rarr; <a href="string.html">string</a></code></td><td><span class="funcdesc"><p>Returns a substring of <code>input</code> starting at <code>start_pos</code> (count starts at 1).</p>
 </span></td></tr>
 <tr><td><a name="substring"></a><code>substring(input: <a href="string.html">string</a>, start_pos: <a href="int.html">int</a>, length: <a href="int.html">int</a>) &rarr; <a href="string.html">string</a></code></td><td><span class="funcdesc"><p>Returns a substring of <code>input</code> starting at <code>start_pos</code> (count starts at 1) and including up to <code>length</code> characters.</p>
+</span></td></tr>
+<tr><td><a name="substring"></a><code>substring(input: varbit, start_pos: <a href="int.html">int</a>) &rarr; varbit</code></td><td><span class="funcdesc"><p>Returns a bit subarray of <code>input</code> starting at <code>start_pos</code> (count starts at 1).</p>
+</span></td></tr>
+<tr><td><a name="substring"></a><code>substring(input: varbit, start_pos: <a href="int.html">int</a>, length: <a href="int.html">int</a>) &rarr; varbit</code></td><td><span class="funcdesc"><p>Returns a bit subarray of <code>input</code> starting at <code>start_pos</code> (count starts at 1) and including up to <code>length</code> characters.</p>
 </span></td></tr>
 <tr><td><a name="to_english"></a><code>to_english(val: <a href="int.html">int</a>) &rarr; <a href="string.html">string</a></code></td><td><span class="funcdesc"><p>This function enunciates the value of its argument using English cardinals.</p>
 </span></td></tr>

--- a/pkg/sql/logictest/testdata/logic_test/builtin_function
+++ b/pkg/sql/logictest/testdata/logic_test/builtin_function
@@ -224,6 +224,86 @@ SELECT substring('f(oabaroob' from '\(o(.)b' for '+')
 query error unknown signature: substring\(\)
 SELECT substring()
 
+# Adding testcases for substring against bit array.
+
+query TTT
+SELECT substring(B'11110000', 0), substring(B'11110000', -1), substring(B'11110000', 5)
+----
+11110000 11110000 0000
+
+query TTT
+SELECT substring(B'11110000', 8), substring(B'11110000', 10), substring(B'', 0)
+----
+0  ·  ·
+
+query TTT
+SELECT substring('11100011'::bit(8), 4), substring('11100011'::bit(6), 4), substring(B'', 0, 1)
+----
+00011 000  ·
+
+query TTT
+SELECT substring(B'11110000', 0, 4), substring(B'11110000', -1, 4), substring(B'11110000', 5, 10)
+----
+111 11 0000
+
+query TTT
+SELECT substring(B'11110000', 8, 1), substring(B'11110000', 8, 0), substring(B'11110000', 10, 5)
+----
+0  ·  ·
+
+query TT
+SELECT substring('11100011'::bit(10), 4, 10), substring('11100011'::bit(8), 1, 8)
+----
+0001100 11100011
+
+query TTT
+SELECT substring(B'10001000' FOR 4 FROM 0), substring(B'10001000' FROM 0 FOR 4), substring(B'10001000' FOR 4)
+----
+100 100 1000
+
+query error substring\(\): negative bit subarray length -1 not allowed
+SELECT substring('11100011'::bit(10), 4, -1)
+
+# Adding testcases for substring against byte array.
+
+query TT
+SELECT substring(b'abc', 0), substring(b'\x61\x62\x63', -1)
+----
+abc abc
+
+query TT
+SELECT substring(b'abc', 3), substring(b'abc', 5)
+----
+c  ·
+
+query T
+SELECT substring('abc'::bytea, 0)
+----
+abc
+
+query TT
+SELECT substring(b'\x61\x62\x63', 0, 4), substring(b'abc', -1, 4)
+----
+abc ab
+
+query TTT
+SELECT substring(b'abc', 3, 1), substring(b'abc', 3, 0), substring(b'abc', 4, 3)
+----
+c  ·  ·
+
+query T
+SELECT substring('abc'::bytea, 0, 4)
+----
+abc
+
+query TTT
+SELECT substring(b'abc' FOR 3 FROM 1), substring(b'abc' FROM 1 FOR 3), substring(b'abc' FOR 3)
+----
+abc abc abc
+
+query error substring\(\): negative byte subarray length -1 not allowed
+SELECT substring('11100011'::bytea, 4, -1)
+
 query error unknown signature: concat_ws\(\)
 SELECT concat_ws()
 

--- a/pkg/sql/logictest/testdata/logic_test/pg_catalog
+++ b/pkg/sql/logictest/testdata/logic_test/pg_catalog
@@ -1278,6 +1278,10 @@ substring  1307062959    NULL      0        NULL     NULL     0
 substring  1307062959    NULL      0        NULL     NULL     0
 substring  1307062959    NULL      0        NULL     NULL     0
 substring  1307062959    NULL      0        NULL     NULL     0
+substring  1307062959    NULL      0        NULL     NULL     0
+substring  1307062959    NULL      0        NULL     NULL     0
+substring  1307062959    NULL      0        NULL     NULL     0
+substring  1307062959    NULL      0        NULL     NULL     0
 
 query TTBBBB colnames
 SELECT proname, protransform, proisagg, proiswindow, prosecdef, proleakproof
@@ -1285,6 +1289,10 @@ FROM pg_catalog.pg_proc
 WHERE proname='substring'
 ----
 proname    protransform  proisagg  proiswindow  prosecdef  proleakproof
+substring  NULL          false     false        false      true
+substring  NULL          false     false        false      true
+substring  NULL          false     false        false      true
+substring  NULL          false     false        false      true
 substring  NULL          false     false        false      true
 substring  NULL          false     false        false      true
 substring  NULL          false     false        false      true
@@ -1300,6 +1308,10 @@ substring  false        false      NULL         NULL
 substring  false        false      NULL         NULL
 substring  false        false      NULL         NULL
 substring  false        false      NULL         NULL
+substring  false        false      NULL         NULL
+substring  false        false      NULL         NULL
+substring  false        false      NULL         NULL
+substring  false        false      NULL         NULL
 
 query TIIOTTTT colnames
 SELECT proname, pronargs, pronargdefaults, prorettype, proargtypes, proallargtypes, proargmodes, proargdefaults
@@ -1311,6 +1323,10 @@ substring  2         0                25          25 20        NULL            N
 substring  3         0                25          25 20 20     NULL            NULL         NULL
 substring  2         0                25          25 25        NULL            NULL         NULL
 substring  3         0                25          25 25 25     NULL            NULL         NULL
+substring  2         0                1562        1562 20      NULL            NULL         NULL
+substring  3         0                1562        1562 20 20   NULL            NULL         NULL
+substring  2         0                17          17 20        NULL            NULL         NULL
+substring  3         0                17          17 20 20     NULL            NULL         NULL
 
 query TTTTTT colnames
 SELECT proname, protrftypes, prosrc, probin, proconfig, proacl
@@ -1318,6 +1334,10 @@ FROM pg_catalog.pg_proc
 WHERE proname='substring'
 ----
 proname    protrftypes  prosrc     probin  proconfig  proacl
+substring  NULL         substring  NULL    NULL       NULL
+substring  NULL         substring  NULL    NULL       NULL
+substring  NULL         substring  NULL    NULL       NULL
+substring  NULL         substring  NULL    NULL       NULL
 substring  NULL         substring  NULL    NULL       NULL
 substring  NULL         substring  NULL    NULL       NULL
 substring  NULL         substring  NULL    NULL       NULL

--- a/pkg/sql/sem/builtins/builtins.go
+++ b/pkg/sql/sem/builtins/builtins.go
@@ -3893,17 +3893,8 @@ var substringImpls = makeBuiltin(tree.FunctionProperties{Category: categoryStrin
 		},
 		ReturnType: tree.FixedReturnType(types.String),
 		Fn: func(_ *tree.EvalContext, args tree.Datums) (tree.Datum, error) {
-			runes := []rune(string(tree.MustBeDString(args[0])))
-			// SQL strings are 1-indexed.
-			start := int(tree.MustBeDInt(args[1])) - 1
-
-			if start < 0 {
-				start = 0
-			} else if start > len(runes) {
-				start = len(runes)
-			}
-
-			return tree.NewDString(string(runes[start:])), nil
+			substring := getSubstringFromIndex(string(tree.MustBeDString(args[0])), int(tree.MustBeDInt(args[1])))
+			return tree.NewDString(substring), nil
 		},
 		Info: "Returns a substring of `input` starting at `start_pos` (count starts at 1).",
 	},
@@ -3916,33 +3907,15 @@ var substringImpls = makeBuiltin(tree.FunctionProperties{Category: categoryStrin
 		SpecializedVecBuiltin: tree.SubstringStringIntInt,
 		ReturnType:            tree.FixedReturnType(types.String),
 		Fn: func(_ *tree.EvalContext, args tree.Datums) (tree.Datum, error) {
-			runes := []rune(string(tree.MustBeDString(args[0])))
-			// SQL strings are 1-indexed.
-			start := int(tree.MustBeDInt(args[1])) - 1
+			str := string(tree.MustBeDString(args[0]))
+			start := int(tree.MustBeDInt(args[1]))
 			length := int(tree.MustBeDInt(args[2]))
 
-			if length < 0 {
-				return nil, pgerror.Newf(
-					pgcode.InvalidParameterValue, "negative substring length %d not allowed", length)
+			substring, err := getSubstringFromIndexOfLength(str, "substring", start, length)
+			if err != nil {
+				return nil, err
 			}
-
-			end := start + length
-			// Check for integer overflow.
-			if end < start {
-				end = len(runes)
-			} else if end < 0 {
-				end = 0
-			} else if end > len(runes) {
-				end = len(runes)
-			}
-
-			if start < 0 {
-				start = 0
-			} else if start > len(runes) {
-				start = len(runes)
-			}
-
-			return tree.NewDString(string(runes[start:end])), nil
+			return tree.NewDString(substring), nil
 		},
 		Info: "Returns a substring of `input` starting at `start_pos` (count starts at 1) and " +
 			"including up to `length` characters.",
@@ -3976,7 +3949,121 @@ var substringImpls = makeBuiltin(tree.FunctionProperties{Category: categoryStrin
 		Info: "Returns a substring of `input` that matches the regular expression `regex` using " +
 			"`escape_char` as your escape character instead of `\\`.",
 	},
+	tree.Overload{
+		Types: tree.ArgTypes{
+			{"input", types.VarBit},
+			{"start_pos", types.Int},
+		},
+		ReturnType: tree.FixedReturnType(types.VarBit),
+		Fn: func(_ *tree.EvalContext, args tree.Datums) (tree.Datum, error) {
+			bitString := tree.MustBeDBitArray(args[0])
+			start := int(tree.MustBeDInt(args[1]))
+			substring := getSubstringFromIndex(bitString.BitArray.String(), start)
+			return tree.ParseDBitArray(substring)
+		},
+		Info: "Returns a bit subarray of `input` starting at `start_pos` (count starts at 1).",
+	},
+	tree.Overload{
+		Types: tree.ArgTypes{
+			{"input", types.VarBit},
+			{"start_pos", types.Int},
+			{"length", types.Int},
+		},
+		ReturnType: tree.FixedReturnType(types.VarBit),
+		Fn: func(_ *tree.EvalContext, args tree.Datums) (tree.Datum, error) {
+			bitString := tree.MustBeDBitArray(args[0])
+			start := int(tree.MustBeDInt(args[1]))
+			length := int(tree.MustBeDInt(args[2]))
+
+			substring, err := getSubstringFromIndexOfLength(bitString.BitArray.String(), "bit subarray", start, length)
+			if err != nil {
+				return nil, err
+			}
+			return tree.ParseDBitArray(substring)
+		},
+		Info: "Returns a bit subarray of `input` starting at `start_pos` (count starts at 1) and " +
+			"including up to `length` characters.",
+	},
+	tree.Overload{
+		Types: tree.ArgTypes{
+			{"input", types.Bytes},
+			{"start_pos", types.Int},
+		},
+		ReturnType: tree.FixedReturnType(types.Bytes),
+		Fn: func(_ *tree.EvalContext, args tree.Datums) (tree.Datum, error) {
+			byteString := string(*args[0].(*tree.DBytes))
+			start := int(tree.MustBeDInt(args[1]))
+			substring := getSubstringFromIndex(byteString, start)
+			return tree.ParseDByte(substring)
+		},
+		Info: "Returns a byte subarray of `input` starting at `start_pos` (count starts at 1).",
+	},
+	tree.Overload{
+		Types: tree.ArgTypes{
+			{"input", types.Bytes},
+			{"start_pos", types.Int},
+			{"length", types.Int},
+		},
+		ReturnType: tree.FixedReturnType(types.Bytes),
+		Fn: func(_ *tree.EvalContext, args tree.Datums) (tree.Datum, error) {
+			byteString := string(*args[0].(*tree.DBytes))
+			start := int(tree.MustBeDInt(args[1]))
+			length := int(tree.MustBeDInt(args[2]))
+
+			substring, err := getSubstringFromIndexOfLength(byteString, "byte subarray", start, length)
+			if err != nil {
+				return nil, err
+			}
+			return tree.ParseDByte(substring)
+		},
+		Info: "Returns a byte subarray of `input` starting at `start_pos` (count starts at 1) and " +
+			"including up to `length` characters.",
+	},
 )
+
+// Returns a substring of given string starting at given position.
+func getSubstringFromIndex(str string, start int) string {
+	runes := []rune(str)
+	// SQL strings are 1-indexed.
+	start--
+
+	if start < 0 {
+		start = 0
+	} else if start > len(runes) {
+		start = len(runes)
+	}
+	return string(runes[start:])
+}
+
+// Returns a substring of given string starting at given position and
+// include upto certain length.
+func getSubstringFromIndexOfLength(str, errMsg string, start, length int) (string, error) {
+	runes := []rune(str)
+	// SQL strings are 1-indexed.
+	start--
+
+	if length < 0 {
+		return "", pgerror.Newf(
+			pgcode.InvalidParameterValue, "negative %s length %d not allowed", errMsg, length)
+	}
+
+	end := start + length
+	// Check for integer overflow.
+	if end < start {
+		end = len(runes)
+	} else if end < 0 {
+		end = 0
+	} else if end > len(runes) {
+		end = len(runes)
+	}
+
+	if start < 0 {
+		start = 0
+	} else if start > len(runes) {
+		start = len(runes)
+	}
+	return string(runes[start:end]), nil
+}
 
 var uuidV4Impl = makeBuiltin(
 	tree.FunctionProperties{

